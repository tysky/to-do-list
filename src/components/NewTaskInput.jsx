import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import * as actionCreators from '../actions';
import Alert from './Alert';

const mapStateToProps = ({ newTaskInputText, taskSendingState, tasks }) => {
  const props = {
    newTaskInputText,
    taskSendingState,
    tasksLength: Object.values(tasks).length,
  };
  return props;
};

class NewTaskInput extends React.Component {
  handleInputText = (e) => {
    const { updateNewTaskText } = this.props;
    updateNewTaskText({ text: e.target.value });
  }

  addTask = (e) => {
    e.preventDefault();
    const { addTask, newTaskInputText, tasksLength } = this.props;
    addTask({ text: newTaskInputText, taskIndex: tasksLength + 1 });
  }

  render() {
    const { newTaskInputText, taskSendingState } = this.props;
    const requested = taskSendingState === 'requested';
    const failed = taskSendingState === 'failed';
    return (
      <>
        <form onSubmit={this.addTask}>
          <label htmlFor="new-task">
          Add Task
            <input
              id="new-task"
              className="addTaskInput"
              type="text"
              required
              value={newTaskInputText}
              onChange={this.handleInputText}
              disabled={requested}
              autoComplete="off"
            />
          </label>
          <button type="submit" disabled={requested}>Add</button>
        </form>
        {failed && <Alert message="Oh snap! Error while adding task. Try again." /> }
      </>
    );
  }
}

NewTaskInput.propTypes = {
  addTask: PropTypes.func.isRequired,
  newTaskInputText: PropTypes.string.isRequired,
  updateNewTaskText: PropTypes.func.isRequired,
  taskSendingState: PropTypes.string.isRequired,
  tasksLength: PropTypes.number.isRequired,
};

export default connect(
  mapStateToProps,
  actionCreators,
)(NewTaskInput);
