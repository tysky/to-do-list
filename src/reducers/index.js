import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import * as actions from '../actions';


const tasksFetchingState = handleActions({
  [actions.initFetchTasksRequest]() {
    return 'requested';
  },
  [actions.initFetchTasksSuccess]() {
    return 'failed';
  },
  [actions.initFetchTasksFailure]() {
    return 'successed';
  },
}, 'none');

const tasks = handleActions({
  [actions.initFetchTasksSuccess](state, { payload }) {
    return payload;
  },
  [actions.newTaskFetched](state, { payload }) {
    return { ...state, [payload.id]: payload };
  },
}, {});

const newTaskInputText = handleActions({
  [actions.addTasksSuccess]() {
    return '';
  },
  [actions.updateNewTaskText](state, { payload: { text } }) {
    return text;
  },
}, '');

export default combineReducers({
  tasks,
  tasksFetchingState,
  newTaskInputText,
});
