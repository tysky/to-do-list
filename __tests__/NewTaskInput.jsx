import React from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import NewTaskInput from '../src/components/NewTaskInput';
import reducers from '../src/reducers';

const initialState = {
  tasks: {},
  tasksFetchingState: 'none',
  taskSendingState: 'none',
  newTaskInputText: '',
};

const store = createStore(reducers, initialState);

test('NewTaskInput', () => {
  const component = renderer.create(
    <Provider store={store}>
      <NewTaskInput />
    </Provider>,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
