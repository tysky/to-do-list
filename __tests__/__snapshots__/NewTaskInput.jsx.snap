// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTaskInput 1`] = `
<form
  onSubmit={[Function]}
>
  <label
    htmlFor="new-task"
  >
    Add Task
    <input
      autoComplete="off"
      className="addTaskInput"
      disabled={false}
      id="new-task"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </label>
  <button
    disabled={false}
    type="submit"
  >
    Add
  </button>
</form>
`;
